kdf.js

```diff
@@ -0,0 +1,181 @@
+const isBuffer = require('is-buffer')
+
+/* eslint-disable camelcase */
+const {
+  crypto_kdf_CONTEXTBYTES,
+  crypto_kdf_KEYBYTES,
+
+  crypto_kdf_derive_from_key,
+  crypto_kdf_keygen,
+} = require('./sodium')
+
+/**
+ * Generates a master key.
+ *
+ * @public
+ * @param {?(Buffer)} [key]
+ * @returns {Buffer}
+ * @throws TypeError
+ */
+function keygen(key) {
+  if (key && false === isBuffer(key)) {
+    throw new TypeError('kdf.keygen: Expecting key to be a buffer.')
+  }
+
+  if (undefined === key) {
+    key = Buffer.allocUnsafe(crypto_kdf_KEYBYTES)
+  }
+
+  if (crypto_kdf_KEYBYTES !== key.length) {
+    throw new TypeError(`kdf.keygen: Invalid key length: ${key.length}`)
+  }
+
+  crypto_kdf_keygen(key)
+  return key
+}
+
+/**
+ * Initializes key derivation.
+ *
+ * @public
+ * @param {Buffer} buffer
+ * @param {Buffer} key
+ * @return {Object}
+ * @throws TypeError
+ */
+function init(buffer, key) {
```

Comment:
We should not need to supply a buffer. This function requires the caller to supply a buffer or a falsy value like `null` to get the "default" effect. The optional buffer should be the second argument.